compiler/src/iree/compiler/Codegen/Common/WorkgroupSpecializationPass.cpp

```diff
@@ -0,0 +1,260 @@
+  SmallVector<AffineMinOp> minOps;
+
+  for (Operation *user : forOp.getInductionVar().getUsers()) {
+    if (isBoundedTileSizeOp(user, tileSize)) {
+      minOps.push_back(cast<AffineMinOp>(user));
+    }
+  }
+
+  return minOps;
+}
+
+// Specialize the tiled distribution loops with the main tile sizes.
+//
+// Input: A vector of LoopTilingAndDistributionInfo
+//
+// Transformed output
+//   cond = (boundedTileSizeY != TileX) && (boundedTileSizeX != TileY) && ...
+//   scf.if cond
+//     distribution loops with static shapes with the tile size
+//   else
+//     distribution loops with dynamic shapes with the tile size
+//
+// Steps:
+//   1. bail out if the loop bounds are already a multiple of the tile size.
+//   2. create a condition for scf.if
+//   3. clone the nested loops in the else block.
+//   4. update the bounded size ops of the original loop nest with the constant
+//      tile size.
+//   5. clone the updated loop nest in the then block.
+static void specializeDistributionLoops(
+    SmallVector<LoopTilingAndDistributionInfo> &infoVec) {
+  // get the scf for loop nests and their tile size.
+  SmallVector<scf::ForOp> distLoops;
+  SmallVector<int64_t> tileSizes;
+  SmallVector<AffineMinOp> minSizeOps;
+
+  // Distribution info vector has the innermost loop at index 0. Let's reverse
+  // it to make the outermost loop positioned at index 0, and collect forOps,
+  // tile sizes, and the bounded tile size ops.
+  for (auto &info : reverse(infoVec)) {
+    auto forOp = cast<scf::ForOp>(info.loop);
+    distLoops.push_back(forOp);
+    if (!info.tileSize) {
+      return;
+    }
+    int64_t tileSize = *info.tileSize;
+    tileSizes.push_back(tileSize);
+    SmallVector<AffineMinOp> boundedTileSizesOps =
+        getBoundedTileSizeOps(forOp, tileSize);
+    // Supposed to see only one affine min for the bounded size.
+    if (boundedTileSizesOps.size() != 1) {
+      return;
+    }
+    minSizeOps.push_back(boundedTileSizesOps[0]);
+  }
+
+  // Check the eligilbility. Unsupported cases are:
+  //   1. Dynamic cases
+  //   2. UB < Tile size
+  //   3. UB == a multiple of the tile size
+  bool isAlreadyMultiple = true;
+  for (unsigned i = 0, e = distLoops.size(); i != e; ++i) {
+    scf::ForOp loop = distLoops[i];
+    auto ubOp = loop.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    if (!ubOp) {
+      // TODO: handle dynamic cases by adding more code to check the
+      // conditions.
+      return;
+    }
+    int64_t ub = ubOp.value();
+    int64_t tileSize = tileSizes[i];
+    if (ub % tileSize != 0) {
+      isAlreadyMultiple = false;
+    }
+    if (ub < tileSize) {
+      return;
+    }
+  }
+
+  if (isAlreadyMultiple) return;
+
+  scf::ForOp outermostLoop = distLoops[0];
+  auto loc = outermostLoop.getLoc();
+  OpBuilder builder(outermostLoop->getContext());
+  PatternRewriter::InsertionGuard guard(builder);
+  builder.setInsertionPoint(outermostLoop);
```

Comment:
This doesn't sound right, the `if` should be in the inner most loop. Why would we insert it above the outer most loop? Those loops are iterating through the tiles. It sounds like you are assuming the loop will run only once?

This also means you wouldn't need to clone the affine_min operations (which I don't think you need)

Reply:
It should be fine. It doesn't make a huge difference in inserting the if/else outside or inside. I specialized the whole distribution loop by treating it as a unit. Since it clones the whole nested loop, there is no assumption that the loop runs only once or not. The concept of the specialization is to specialize as top as possible in the dispatch and I thought inserting if/else around the distribution loop is the way.

Cloning affine.min seems to be necessary even the innermost loop approach, but it depends on how we implement the cloning and replace the affine.min ops in the then block.